RFC process proposal

aturon opened the pull request:
As the Bytecode Alliance grows, we need more formalized ways of communicating about and reaching consensus on major changes to core projects. This document proposes to adapt ideas from Rust’s [RFC](https://github.com/rust-lang/rfcs/) and [MCP](https://forge.rust-lang.org/compiler/mcp.html) processes to the Bytecode Alliance context.

[Rendered proposal text](https://github.com/aturon/ba-rfcs/blob/rfc-process/accepted/rfc-process.md)

joshtriplett commented on accepted/rfc-process.md:
```diff
@@ -0,0 +1,109 @@
+
+# Summary
+
+As the Bytecode Alliance grows, we need more formalized ways of communicating about and reaching consensus on major changes to core projects. This document proposes to adapt ideas from Rust’s [RFC](https://github.com/rust-lang/rfcs/) and [MCP](https://forge.rust-lang.org/compiler/mcp.html) processes to the Bytecode Alliance context.
```
There's a formal proposal for the MCP process at https://github.com/rust-lang/rfcs/pull/2936/ ; I would suggest linking to that.

Also, please expand "RFC" and "MCP" on first use.

aturon replied:
I'll expand the acronyms :+1:

Re: the MCP link, I'd prefer to keep it pointed at the official documentation on the Forge, which post-dates the compiler team RFC (which was accepted a while back).

joshtriplett commented on accepted/rfc-process.md:
```diff
@@ -0,0 +1,109 @@
+
+Each core BA project has a formal set of **stakeholders**. These are individuals, organized into groups by project and/or member organization. Stakeholders can block proposals, and conversely having explicit sign-off from at least one individual within each stakeholder group is a sufficient (but not necessary) condition for immediately accepting a proposal.
+
+The process for determining core BA projects and their stakeholder set will ultimately be defined by the Technical Steering Committee, once it is in place. Until then, the current BA Steering Committee will be responsible for creating a provisional stakeholder arrangement, as well as deciding whether to accept this RFC.
```
Do we expect the stakeholder set to be completely unbounded in number? If so, can we say that explicitly? If not, how do we handle that?

aturon replied:
I'm not sure what you have in mind about "unbounded in number" here. I don't foresee a hard limit on the stakeholder count, but rather a policy around stakeholders that naturally results in a reasonable upper bound in practice. For example, a likely criterion for formal stakeholdership is being a major contributor to the project itself, a number which is likely to be self-limiting for other reasons.

All that said, I'm also wary of trying to specify too much about the stakeholder determination in this RFC, preferring to treat it as a modular detail ultimately owned by the TSC.

joshtriplett commented on accepted/rfc-process.md:
```diff
@@ -0,0 +1,109 @@
+
+# Proposal
+
+The design of this RFC process draws ideas from Rust’s [RFC](https://github.com/rust-lang/rfcs/) and [MCP](https://forge.rust-lang.org/compiler/mcp.html) processes, adapting to the BA and trying to keep things lightweight.
```
Same comment as in the summary about where MCP links to.

joshtriplett commented on accepted/rfc-process.md:
```diff
@@ -0,0 +1,109 @@
+
+*   We have a dedicated bytecodealliance/rfcs repo that houses _all_ RFCs for core BA projects, much like Rust’s rfcs repo that is shared between all Rust teams. 
+*   The rfcs repo will be structured similarly to the one in Rust:
+    *   A template markdown file laying out the format of RFCs, like [Rust’s](https://github.com/rust-lang/rfcs/blob/master/0000-template.md) but simplified.
```
I would suggest that our process should take inspiration from the current direction of Rust processes, and start with a more MCP-like template (focused on the problem statement and not just the proposal). There may also be value in adopting the processes for assigning a *liaison*, and having the option of designating a *project group* to shepherd complex proposals.

aturon replied:
So, it seems like a point of confusion here is that there are _multiple_ notions of "MCP" in flight in the Rust community. There's the existing one for the Compiler team, which doesn't even have a problem statement section, and the one in the works for the language team.

I think what'd be helpful here is to get more concrete: is there something about the template, or some other part of the process, that is underemphasizing writing about the motivation?

Note also that there's an explicit separate template for early-stage ("draft") RFCs that pretty closely resembles the proposed Rust Lang Team MCP template.

re: liaisons and project groups, my feeling is that we are probably too early on in the BA for those ideas to apply well. (In particular, the total group of people writing or consuming RFCs is likely to be quite small at the outset, compared to where Rust was even back at the 1.0 days). I suspect that we'll want to evolve the RFC process over time as the unique needs of the BA become more clear.

joshtriplett commented on accepted/rfc-process.md:
```diff
@@ -0,0 +1,109 @@
+    *   A template markdown file laying out the format of RFCs, like [Rust’s](https://github.com/rust-lang/rfcs/blob/master/0000-template.md) but simplified.
+    *   A subdirectory holding the text of all accepted RFCs, like [Rust’s](https://github.com/rust-lang/rfcs/tree/master/text).
+*   New RFCs are submitted via pull request, and both technical and process discussion happens via the comment thread.
+*   The RFC is tagged with a **project label**, corresponding to the BA project it targets. This tagging informs tooling about the relevant stakeholder set.
```
```suggestion
*   The RFC is tagged with **project labels**, corresponding to the BA project or projects it targets. This tagging informs tooling about the relevant stakeholder set.
```

joshtriplett commented on accepted/rfc-process.md:
```diff
@@ -0,0 +1,109 @@
+*   In response to the motion to finalize, a bot will post a special comment with a **stakeholder checklist**. 
+    *   This list includes the GitHub handle for each individual stakeholder, organized into stakeholder groups. 
+    *   The individual who filed the motion to finalize is automatically checked off.
+*   Once _any_ stakeholder from a _different_ group has signed off, the RFC will move into a 10 day **final comment period** (FCP), long enough to ensure that other stakeholders have at least a full business week to respond.
```
I like this very much; it's a good balance between "FCP starts right away" and "FCP doesn't start until almost everyone has signed off". It seems like a good fit given the other changes we've made to the process.

joshtriplett commented on accepted/rfc-process.md:
```diff
@@ -0,0 +1,109 @@
+*   When discussion has stabilized around the main points of contention, any stakeholder can make a **motion to finalize**, using a special comment syntax understood by tooling. This motion comes with a disposition: merge or close.
+*   In response to the motion to finalize, a bot will post a special comment with a **stakeholder checklist**. 
+    *   This list includes the GitHub handle for each individual stakeholder, organized into stakeholder groups. 
+    *   The individual who filed the motion to finalize is automatically checked off.
```
You might mention, somewhere in this proposal, that we may be able to use Rust's rfcbot for this. (In general, I think it's important for proposals that assume certain technological support to be clear on whether that support is already mostly written or would need to be written.)

joshtriplett commented on accepted/rfc-process.md:
```diff
@@ -0,0 +1,109 @@
+
+### Making a decision: merge or close
+
+*   When discussion has stabilized around the main points of contention, any stakeholder can make a **motion to finalize**, using a special comment syntax understood by tooling. This motion comes with a disposition: merge or close.
```
I do think these two dispositions suffice, but given that, we should say something like "closing an RFC does not imply a refusal to reconsider it in the future if new information arises; a motion to postpone will also use the disposition 'close', along with an explanation.".

joshtriplett commented on accepted/rfc-process.md:
```diff
@@ -0,0 +1,109 @@
+*   **Motivation**. What problem does the RFC solve?
+*   **Proposal**. The meat of the RFC.
+*   **Rationale and alternatives**. A discussion of tradeoffs: why was the proposal chosen, rather than alternatives?
+*   **Open questions**. Often an RFC is initially created with a broad proposal but some gaps that need community input to fill in.
```
As stated above, I think there's value in adopting something closer to the MCP process (focused on the problem rather than a specific propsal) rather than the RFC process, and letting a full RFC come later if and when needed. We may want to have both options. That way, we have a way to charter a group to solve a given problem for which the solution is not predetermined.

joshtriplett commented on accepted/rfc-process.md:
```diff
@@ -0,0 +1,109 @@
+
+*   **Motivation**. What problem are you ultimately hoping to solve?
+*   **Proposal sketch**. At least a short sketch of a possible approach. Beginning discussion without _any_ proposal tends to be unproductive.
+*   **Open questions**. This section is especially important for draft RFCs: it’s where you highlight the key issues you are hoping that discussion will address.
```
Ah, I see. Is this intended to be closer to the MCP?

I think there'd be value in explaining this first, suggesting that almost all proposals should start with this, and then explaining the format of a full RFC second.

joshtriplett commented on accepted/rfc-process.md:
```diff
@@ -0,0 +1,109 @@
+    *   Major architectural changes
+    *   Major new features
+    *   Simple changes that have significant downstream impact
+    *   Changes that could affect guarantees or level of support, e.g. removing support for a target platform
```
Or adding support for a platform with the expectation that it must be maintained going forward.

aturon commented:
I've just pushed a number of commits addressing the first round of reviews. I took the liberty of marking review comments "resolved" when they were straightforwardly addressed by these news commits. The remaining comments need further discussion.

The commits also include a new option question regarding indefinite blocking of proposals.

aturon commented:
Thanks @tschneidereit! I've addressed your review comments with a [new commit](https://github.com/bytecodealliance/rfcs/pull/1/commits/d1a7db726f0a418cf317e9cd3ce426f4be829d2c).

tschneidereit commented on accepted/rfc-process.md:
```diff
@@ -115,10 +115,10 @@ The Rust RFC process has successfully governed the development of the language f
 
-* As written, an individual stakeholder can block acceptance of an RFC indefinitely. In the Rust community, such an arrangement has sometimes caused problems, leading to a secondary process that kicks in after a waiting period, and requires an _additional_ stakeholder to approve continued blocking. Should we consider building in such a mechanism now, or add it to the process only if it becomes ne ...
\ No newline at end of file
+* As written, an individual stakeholder can block acceptance of an RFC indefinitely. In the Rust community, such an arrangement has sometimes caused problems, leading to a secondary process that kicks in after a waiting period, and requires an _additional_ stakeholder to approve continued blocking. Should we consider building in such a mechanism now, or add it to the process only if it becomes ne ...
```
Let's discuss this on Monday, but my take is that we should see how things work for now, and apply remedies if and when we see symptoms of something going wrong.

fitzgen commented:
Is there anything still outstanding for this RFC process PR? Is it ready to merge?

tschneidereit commented:
We effectively started using this process, with a few new RFCs currently pending, so we should indeed merge this. The process for getting an RFC accepted as [described in this proposal itself](https://github.com/aturon/ba-rfcs/blob/rfc-process/accepted/rfc-process.md#structure-and-workflow) includes mention of some tooling that wee don't yet have in place, but all that'd do can instead happen manually, so I'll do that here.

An additional piece we don't yet have in place are the [stakeholder groups](https://github.com/aturon/ba-rfcs/blob/rfc-process/accepted/rfc-process.md#stakeholders) as described in this RFC. That, too, I think we can handle in a more manual way for now, by including everyone who has commit access to affected project repositories, grouped by affiliation. For this PR, I'll aim for completeness, and err on the side of tagging too many people :)

tschneidereit commented:
# Motion to finalize with a disposition to merge

I'd like to propose that we accept this RFC by merging this PR. I'm labeling the PR with the `motion-to-finalize` and `disposition-merge` labels. As described in the RFC, I'm tagging stakeholders of affected projects, which right now are the core collaborative projects Cranelift and Wasmtime. Given the fundamental nature of this RFC, I'll go even more broad in whom I'm tagging for sign-off, by including regular contributors (very arbitrarily defined as having contributed 5+ commits with 100+ lines changed in the last 3 months), regardless of whether they're committers or not, in addition to contributors to this particular RFC, and representatives of the BA members.

## Stakeholders sign-off

### Red Hat
- [ ] @npmccallum

### Mozilla
- [ ] @yurydelendik 
- [ ] @sdetar 
- [ ] @lars-t-hansen 
- [ ] @julian-seward1 
- [ ] @eqrion 

### Intel
- [ ] @mingqiusun 
- [ ] @jlb6740 
- [ ] @abrown 

### IBM
- [ ] @uweigand

### Fastly
- [x] @tschneidereit 
- [ ] @sunfishcode 
- [ ] @pchickey 
- [ ] @linclark 
- [ ] @fitzgen 
- [ ] @cfallin 
- [ ] @aturon 
- [ ] @alexcrichton 
- [ ] @acfoltzer 

### DFINITY
- [ ] @ggreif

### AssemblyScript
- [ ] @MaxGraey

### Arm
- [ ] @akirilov-arm
- [ ] @jgouly

### Unaffiliated
- [ ] @joshtriplett 
- [ ] @bnjbvr 
- [ ] @bjorn3

## Next steps
With this comment, the RFC enters the `motion-to-finalize` phase. As soon as a member of any second group signs off, it'll enter the `final-comment-period` of up to 10 days.

If during that time a member of a key stakeholder group raises a blocking concern, the FCP will be paused until the concern has been resolved. Note that for the sake of that part of the process, only committers and member representatives can raise blocking concerns as a formal matter, but we'd aim to resolve any concerns, regardless of who raises them, as best as possible.

Once the FCP has ended, or at least one member of each stakeholder group has signed off — whichever comes first — this PR can be merged and the FCP accepted.

lars-t-hansen commented:
As a purely administrative matter, Mozilla is closed all of next week, so the timing is not fabulous for those of us who have not seen this previously.

tschneidereit commented:
> As a purely administrative matter, Mozilla is closed all of next week, so the timing is not fabulous for those of us who have not seen this previously.

Ah, thank you for raising that. Would keeping the PR open until at least the end of the week after next address that issue?

lars-t-hansen commented:
Yeah, should be fine.

ggreif commented on accepted/rfc-process.md:
```diff
@@ -0,0 +1,124 @@
+*   **Summary**. A ~one paragraph overview of the RFC.
+*   **Motivation**. What problem does the RFC solve?
+*   **Proposal**. The meat of the RFC.
+*   **Rationale and alternatives**. A discussion of tradeoffs: why was the proposal chosen, rather than alternatives?
```
Where would risk assessments go? Should a proposal address some pain point, at the cost of potentially introducing other problems, one needs to elaborate the latter, and weigh in on their severity. One could argue that this belongs into the discussion of the proposal, but that thread is not a document.

tschneidereit replied:
The intent was to have risk assessment be covered by this section, but I agree that it's a bit implicit. I see you've signed off by now (thank you!), but do you think it'd be strongly preferable to address this now? To me it seems like we can include this in a future revision after we learn a bit more about the process in practice.

ggreif replied:
No need to get into it right now. Let's wait for a fine-tuning revision, and get some phrasing in then.

tschneidereit replied:
I filed #7 to track this.

joshtriplett commented:
I do think there's value in defining at least the skeleton of further process for stakeholder disagreement now, rather than the first time we need it (when tensions and stakes may be higher); evidence from other communities (such as Rust) suggests we will eventually need it.

That said, the RFC as written looks good to me. Thanks for driving this, @aturon!

lars-t-hansen commented:
@tschneidereit, I'm fine with the proposal but I don't think I have write permissions here and can't record the vote directly.

ggreif reviewed (approved):
LGTM!

yurydelendik commented on accepted/rfc-process.md:
```diff
@@ -0,0 +1,124 @@
+
+# Proposal
+
+The design of this RFC process draws ideas from Rust’s [Request for Comment](https://github.com/rust-lang/rfcs/) (RFC) and [Major Change Proposal](https://forge.rust-lang.org/compiler/mcp.html) (MCP) processes, adapting to the BA and trying to keep things lightweight.
```
Is it expected for this process to deviate from Rust's one? In case of if either process changes, the statements "like Rust's" or "similarly to the one in Rust" might become incorrect. Is it possible to point this proposal to specific version of Rust's RFC process?

tschneidereit replied:
Good point that those processes might change, and that these references might become stale as a result. I think that that's fine though, since over time it'll also become less important to be able to anchor expectations to related processes in other communities. But we could certainly include more stable links here; would you be up for looking them up, @yurydelendik?

tschneidereit commented:
> I do think there's value in defining at least the skeleton of further process for stakeholder disagreement now, rather than the first time we need it (when tensions and stakes may be higher); evidence from other communities (such as Rust) suggests we will eventually need it.

Thank you for raising this, @joshtriplett. I agree that we'll need this, but there's a tradeoff between trying to get all related processes in place and getting this core process formally instated. We should certainly not put it off indefinitely :) I filed #6 to track this.

tschneidereit commented:
# Entering Final Call Period

Since we have sign-off by members of multiple stakeholder groups and no blocking concerns have been raised, I'm moving this RFC to Final Call. That means it'll be merged in 10 days from now, or once the remaining stakeholder groups have signed off on it, provided no blocking concerns are raised in the meantime.

tschneidereit commented:
I received a sign-off from the last missing stakeholder group in private message, so we have full approval, in addition to the 1-day FCP having lapsed :) With that, I'll merge this PR, at which point the RFC process is formally instated.
